Users can mark a bedspace unavailable for bookings

patrickjfl opened the pull request:
# Context

Allows users to void a bedspace, and to view void 'bookings'. See [Trello ticket](https://trello.com/c/aP1llGe6/756-users-can-mark-a-bedspace-as-unavailable-for-bookings) for more information. Much of the work shadows the existing Bookings journey. For now the Voids routes and link within the UI are feature flagged, pending API support.

Also two minor tweaks, one to the .env.example file, and another to the restClient logging.

# Changes in this PR

## Screenshots of UI changes

![Screenshot 2023-02-24 at 16 04 58](https://user-images.githubusercontent.com/70749355/221229330-6b74bce6-5cd9-423f-ad92-70d26ea2fb0e.png)
![Screenshot 2023-02-24 at 16 04 43](https://user-images.githubusercontent.com/70749355/221229324-8d61fa13-057d-4e12-8284-6435335bb57e.png)

### Before

![Screenshot 2023-02-24 at 16 11 34](https://user-images.githubusercontent.com/70749355/221229083-900ecc21-5ebf-4508-af63-212f1a6d5e69.png)

### After

![Screenshot 2023-02-24 at 16 04 35](https://user-images.githubusercontent.com/70749355/221229174-cbe3ec48-c53d-46fa-9894-5d6effd6fac3.png)

# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [ ] Have all changes to any dependencies been deployed to both `preprod` and
    `production` in a backwards compatible way? (This includes our API,
    infrastructure, third-party integrations and any secrets or environment variables)
- [ ] Has a required data migration been included in this change or been done in
    advance?

## Post merge checklist

Once we've merged we now need to release this through to production before
considering it done.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-temporary-accommodation-ui)
and work through the following steps:

- [ ] Has the product manager asked to provide approval for this feature?
  - [ ] Has the product manager approved?
- [ ] Manually approve release to preprod
- [ ] Verify change released to preprod
- [ ] Manually approve release to prod
- [ ] Verify change released to production

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

Commit:
Update .env.example

We replace legacy service name.

Commit:
Fix restClient error handling

We correct restClient error handling that was causing POST and PUT errors to be logged only as PUT errors.

Commit:
Add feature flag for voids button and routes

We hide the ‘Void bedspace’ action on the view bedspace screen, as well as the routes used by the lostBedController, behind a voidsDisabled feature flag.

bryangaledxw commented:
For the Trello link, you can use the feature in Trello to get a bidirectional link between ticket and PR
![image](https://user-images.githubusercontent.com/94137563/221548106-c0561d39-24e1-454b-98fb-a30b8d308283.png)

bryangaledxw commented on server/testutils/factories/newLostBed.ts:
```diff
@@ -3,24 +3,23 @@ import { faker } from '@faker-js/faker/locale/en_GB'
+    endDate: null,
     'endDate-day': endDate.getDate().toString(),
     'endDate-month': endDate.getMonth().toString(),
     'endDate-year': endDate.getFullYear().toString(),
```
My preference would be go the other way here - keep the factory 'honest' wrt the `NewLostBed` type, and what gets sent from the UI to the API, and do any extra work for turn this into what gets sent from the page to the UI in the test

I've run into factories inherited from CAS1 doing this, and got rid of it because it ended up being something of a tripwire

patrickjfl replied:
yep ok

bryangaledxw commented on wiremock/lostBedStubs.ts:
```diff
@@ -15,10 +15,41 @@ lostBeds.push({
+  priority: 99,
+  request: {
+    method: 'GET',
+    urlPathPattern: `/premises/${guidRegex}/lost-beds/${guidRegex}`,
```
It feels a little dodgy given that you'd be sticking the `guidRegex` in as an ID, but it seems static paths do work here. See `applicationStubs.ts`

patrickjfl replied:
ah ok! I wasn't sure, I think different files are doing different things.

bryangaledxw commented on server/components/lostBedInfo.ts:
```diff
@@ -0,0 +1,41 @@
+import { formatLines } from '../utils/viewUtils'
+import { lostBedStatusHtml } from '../utils/lostBedUtils'
+
+export default (lostBed: LostBed): SummaryList['rows'] => {
```
Does this go with a macro and changes in nunjucksSetup.ts?  Similar to `booking-info/macro.njk`/`bookingInfo.ts`? Suggest add those all together in one commit

patrickjfl replied:
done

bryangaledxw commented on server/services/lostBedService.ts:
```diff
@@ -13,11 +13,19 @@ export default class LostBedService {
     const lostBedClient = this.lostBedClientFactory(callConfig)
 
-    const confirmedLostBed = await lostBedClient.create(premisesId, lostBed)
+    const confirmedLostBed = (await lostBedClient.create(premisesId, lostBed)) as LostBed
```
I don't think the cast is needed?

patrickjfl replied:
👍

bryangaledxw commented on server/services/bookingService.test.ts:
```diff
@@ -1,26 +1,35 @@
 import BookingService from './bookingService'
 import BookingClient from '../data/bookingClient'
+import LostBedClient from '../data/lostBedClient'
```
Suggest keep the changes to booking service in their own commit

bryangaledxw commented on server/controllers/temporary-accommodation/manage/lostBedsController.test.ts:
```diff
@@ -0,0 +1,183 @@
+          reason: lostBed.reason.id,
+          bedId: lostBed.bed.id,
+          referenceNumber: lostBed.referenceNumber,
+          serviceName: 'temporary-accommodation',
```
Can you use the newLostBedFactory, and replace this all with something like:
```
...newLostBed,
...DateFormats.convertIsoToDateAndTimeInputs(lostBed.startDate, 'startDate'),
...DateFormats.convertIsoToDateAndTimeInputs(lostBed.endDate, 'endDate'),
```
?

patrickjfl replied:
yes, good shout

bryangaledxw commented on server/controllers/temporary-accommodation/manage/lostBedsController.test.ts:
```diff
@@ -0,0 +1,183 @@
+        expect(lostBedService.createLostBed).toHaveBeenCalledWith(callConfig, premisesId, {
+          ...request.body,
+          startDate: '2022-08-22',
+          endDate: '2022-09-22',
```
... then this can become something like `expect.objectContaining(newLostBed)`

bryangaledxw commented on server/views/temporary-accommodation/components/lost-bed-info/macro.njk:
```diff
@@ -0,0 +1,15 @@
+{% from "govuk/components/summary-list/macro.njk" import govukSummaryList %}
```
For all the templates, add a newline at the end, and suggest add some more whitespace for readability

bryangaledxw commented on server/config.ts:
```diff
@@ -49,6 +49,7 @@ export default {
   staticResourceCacheDuration: 20,
   flags: {
     oasysDisabled: process.env.OASYS_DISABLED || false,
+    voidsDisabled: !['local', 'dev', 'test'].includes(process.env.ENVIRONMENT || 'local'),
```
Suggest pull out `const environment = process.env.ENVIRONMENT || 'local'`, and use that here and on line 54

patrickjfl replied:
good call

bryangaledxw commented on cypress_shared/pages/temporary-accommodation/manage/lostBedEditable.ts:
```diff
@@ -0,0 +1,30 @@
+  }
+
+  protected completeEditableForm(newOrUpdateLostBed: NewLostBed, roomId: string): void {
+    this.getHiddenInputByNameAndSetValue('bedId', roomId)
```
I feel like in principle this shouldn't be necessary? Everything on the page is coming from our own faked API, so we should be able control the contents of the hidden for via that?

patrickjfl replied:
hmmm yeah, not sure what I was trying to achieve there. I guess that was when I was confused about bedIds and roomIds. Removed now anyway 👍

bryangaledxw commented on server/services/lostBedService.ts:
```diff
@@ -13,11 +13,19 @@ export default class LostBedService {
   async createLostBed(callConfig: CallConfig, premisesId: string, lostBed: NewLostBed): Promise<LostBed> {
```
Suggest rename to "create" for consistency

bryangaledxw commented on server/views/temporary-accommodation/lost-beds/new.njk:
```diff
@@ -0,0 +1,29 @@
+        <div class="govuk-grid-column-two-thirds">
+            <form action="{{ paths.lostBeds.create({ premisesId: premises.id, roomId: room.id }) }}" method="post">
+                <input type="hidden" name="_csrf" value="{{ csrfToken }}"/>
+                <input type="hidden" name="bedId" value="{{ room.id }}"/> {% include "./_editable.njk" %}
```
Is this meant to be `value="{{ room.beds[0].id }}"`?

So I understand this correctly, this is so that in the create endpoint, we don't have to fetch the room again *just* to get the bed ID (this is what BookingsController does, but I think your solution here is good)

patrickjfl replied:
ah yes good spot!

Yes, that and to maintain consistency in handling the form data. I think it was @richpjames idea actually, so props to him...

bryangaledxw commented on integration_tests/tests/temporary-accommodation/manage/lostBed.cy.ts:
```diff
@@ -0,0 +1,253 @@
+    // And I fill out the form
+    const lostBed = lostBedFactory.build({
+      bed: bedFactory.build({
+        id: room.id,
```
My instinct is something's gone amiss if we're assigning the room ID to the bed ID. Is this fixed by the change I suggested to `new.njk`?

patrickjfl replied:
yes, good spot

bryangaledxw reviewed (commented):
All looks like really good thorough stuff! So please don't take all my comments as a bad sign!

The only real technical thing I'm worried about is in the integration test - needing to set the hidden field in the integration test, and setting the bed ID to the room ID in the when building a lost bed. I'm hoping this goes away with the change I suggested in new.njk.

Other than that, I'd maybe split the commits a bit differently. I'd really like to see the Lost Bed Info component come in in it's own commit, so the code, the nunjucks template, and the nunjucks setup are self contained. And maybe do a commit each for the changes in BookingService and LostBedService

And I suggest in the controller commit, don't include the route changes, or the changes to show.njk in that. I'd instead put those in with the feature flag commit. So there's one commit adding the controller and its dependencies, and another commit actually wiring it up to the rest of the app

Also, for the unrelated tweaks, I think it's best to have all those first and get them out the way (and I'd put deleting `api.ts` in its own commit)

Last minor things - there's a weird character in the commit message for the lost beds controller commit. And there's a stray period in the title of the integration tests commit

patrickjfl commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Users can mark a bedspace as unavailable for bookings](https://trello.com/c/aP1llGe6/756-users-can-mark-a-bedspace-as-unavailable-for-bookings)

bryangaledxw commented on server/views/temporary-accommodation/lost-beds/show.njk:
```diff
@@ -0,0 +1,37 @@
+    <p>{{ room.name }}</p>
+    <h2>Property address</h2>
+    <p>{{ premises.addressLine1 }}<br/>{{ premises.postcode }}</p>
+  </div>
```
You may well already have this in mind, but let's replace this with the location header component once you've rebased this branch

patrickjfl replied:
yep 👍

Commit:
Add LostBedInfo component

Following the pattern of the BookingInfoComponent, we add a component for rendering details about lost beds.

Commit:
Extend lostBedService

We add a find method to lostBedService and a util to generate html to render Void statuses.

Commit:
Route LostBedsController

We add routing to make controller methods reachable by URL, and render the button users will click to access the ‘Void a bedspace’ journey

Commit:
Delete api.ts

Deletes legacy file.

Commit:
Add LostBedController

We add a lostBedController and nunjucks templates to support new, create and show methods. We add a macro to render lost bed information to the user.

We also update the json error mapping to support 409 conflict errors relating to lost beds.

Commit:
Add integration tests for lost beds journey

bryangaledxw commented on server/@types/shared/models/LostBed.ts:
```diff
@@ -2,18 +2,17 @@
 /* tslint:disable */
 /* eslint-disable */
 
```
Suggest leave this file unlinted, otherwise we'll get spurious changes next time we regenerate types

bryangaledxw commented on server/@types/shared/models/LostBedReason.ts:
```diff
@@ -3,9 +3,8 @@
 /* eslint-disable */
 
 export type LostBedReason = {
```
Suggest leave this file unlinted, otherwise we'll get spurious changes next time we regenerate types

bryangaledxw commented on integration_tests/tests/temporary-accommodation/manage/lostBed.cy.ts:
```diff
@@ -6,7 +6,6 @@ import newLostBedFactory from '../../../../server/testutils/factories/newLostBed
 import BedspaceShowPage from '../../../../cypress_shared/pages/temporary-accommodation/manage/bedspaceShow'
 import Page from '../../../../cypress_shared/pages/page'
 import LostBedNewPage from '../../../../cypress_shared/pages/temporary-accommodation/manage/lostBedNew'
```
Should the changes in this file be with the integration test commit?

bryangaledxw commented on integration_tests/tests/temporary-accommodation/manage/lostBed.cy.ts:
```diff
@@ -163,9 +160,7 @@ context('Lost bed', () => {
-      bed: bedFactory.build({
-        id: room.id,
-      }),
+      bedId: room.id,
```
Should this and the below change be `bedId: room.beds[0].id`?

patrickjfl replied:
yes, good spot

bryangaledxw commented on e2e/tests/lostBed.feature:
```diff
@@ -0,0 +1,18 @@
+        And I attempt to mark a bedspace as void with required details missing
+        Then I should see a list of the problems encountered voiding the bedspace
+
+    
```
Suggest tidy so it ends with a newline

bryangaledxw commented on cypress_shared/pages/temporary-accommodation/manage/bedspaceShow.ts:
```diff
@@ -65,6 +65,7 @@ export default class BedspaceShowPage extends Page {
     cy.get('tr')
       .contains('Void')
       .parent()
+      .parent()
```
I think this should be part of the previous commit?

bryangaledxw commented on e2e/tests/stepDefinitions/manage/lostBed.ts:
```diff
@@ -0,0 +1,56 @@
+import lostBedFactory from '../../../../server/testutils/factories/lostBed'
+import LostBedShowPage from '../../../../cypress_shared/pages/temporary-accommodation/manage/lostBedShow'
+
+Given("I'm marking a bedspace as void", () => {
```
Appreciate I've not been fully consistent, but I think let's standardise on `Given("I mark a bedspace as void")` etc. for how we word these

patrickjfl replied:
ok - as you say there are inconsistencies with other scenario descriptions but we can deal with those later.

bryangaledxw commented on server/controllers/temporary-accommodation/manage/lostBedsController.ts:
```diff
@@ -0,0 +1,89 @@
+        room,
+        lostBed,
+        actions: null,
+        allStatuses,
```
I think this is unused? I can't see that the template does anything with this

patrickjfl replied:
yes I think I was trying to be clever and get ahead of the game for the edit/cancel work but as you say, not necessary here

bryangaledxw commented on server/utils/lostBedUtils.ts:
```diff
@@ -0,0 +1,18 @@
+import type { TemporaryAccommodationLostBed } from '@approved-premises/api'
+
+export const allStatuses: Array<{ name: string; id: TemporaryAccommodationLostBed['status']; tagClass: string }> = [
```
I think right now this isn't used, and the data here gets duplicated in `lostBedInfo.ts`. I'd rather `lostBedStatusHtml` work the same as `formatStatus` elsewhere, and this just take the status ID string, then lookup human-readable string/colour

patrickjfl replied:
generally agree, although the problem with changing lostBedStatusHtml is that it's an unusual case - on the view void booking page we want to render 'Active' or 'Cancelled' to the user, whereas on the view bedspace page, if it's an active void booking we want to render 'Void' to the user

patrickjfl replied:
I've followed the pattern but added an optional parameter and if/else for the view bedspace screen case

bryangaledxw replied:
Is this two functions then? One that does a similar mapping from status ID to HTML as elsewhere. Then a function in `viewUtils.ts` for when we just want to turn some string and some colour into HTML, and isn't void-specific?

bryangaledxw commented on server/utils/lostBedUtils.ts:
```diff
@@ -0,0 +1,18 @@
+  },
+]
+
+export const lostBedStatusHtml = (status: string, colour: string): string => {
```
Heads up that I've renamed similar functions to `statusTag`, to match CAS1 naming of equivalent functions we've pulled over

bryangaledxw commented on server/services/bookingService.ts:
```diff
@@ -50,28 +54,51 @@ export default class BookingService {
 
-    return bookings
-      .filter(booking => booking?.bed.id === bedId)
+    const allReservations = [...bookings, ...lostBeds]
```
Feel free to reject this, but I'm wondering if we can lose the need for `isBooking` - it makes me slightly uneasy that we know what's a void and what's a booking, we then lose that information by putting them in an array together, and we then derive it again by being lucky that they don't share a particular field.

We could do something like -

```
const foo1 = bookings.map(b => {sortingValue: v.arrivalDate, rows: [...]}
const foo2 = voids.map(v => {sortingValue: v.startDate, rows: [...]}

return [...foo1, foo2].sort(...).map(bookingOrVoid => bookingOrVoid.rows)
```

We then don't rely on being able to distinguish between voids and bookings in a roundabout way, and we don't need to do `isBooking` for each row

patrickjfl replied:
yes, good idea. Have implemented

bryangaledxw reviewed (commented):
Good stuff on the E2E tests. A few more minor comments. Main thing is on `lostBedStatusHtml`. I think let's have that work the same as similar functions elsewhere - apologies I missed that first time few. I also want to flag my comment on `lostBedController.test` from my first review, about dodging the need to manually specify the date components, I think maybe that go lost with stuff moving round

Commit:
Extend lostBedClient, update lostBed models and Wiremock

We add find and allLostBedsForPremisesId methods to the lostBedClient. We also update the models and factories to reflect the new structures for LostBed, NewLostBed and LostBedReason. We update Wiremock to stub responses to the lost-beds endpoints.

bryangaledxw commented on server/utils/bookingUtils.test.ts:
```diff
@@ -1,15 +1,15 @@
-import bookingFactory from '../testutils/factories/booking'
-import departureFactory from '../testutils/factories/departure'
-import extensionFactory from '../testutils/factories/extension'
 import {
   bookingActions,
   deriveBookingHistory,
   formatStatus,
   getLatestExtension,
   shortenedOrExtended,
 } from './bookingUtils'
+import bookingFactory from '../testutils/factories/booking'
+import paths from '../paths/temporary-accommodation/manage'
+import arrivalFactory from '../testutils/factories/arrival'
+import departureFactory from '../testutils/factories/departure'
+import extensionFactory from '../testutils/factories/extension'
```
I'd either get rid of this, or move it into a "Sort imports" commit. Also, the commit message still has reference to the type guard util that's no longer here

bryangaledxw commented on server/services/bookingService.ts:
```diff
@@ -50,31 +54,56 @@ export default class BookingService {
-          this.textValue(DateFormats.isoDateToUIDate(booking.arrivalDate, { format: 'short' })),
-          this.textValue(DateFormats.isoDateToUIDate(booking.departureDate, { format: 'short' })),
-          this.htmlValue(formatStatus(booking.status)),
+    const sortedBookings = bookings
```
Is there a better name for `sortedBookings` and `sortedLostBeds`, since these aren't sorted?

patrickjfl replied:
ah yeah, true

bryangaledxw commented on server/utils/lostBedUtils.ts:
```diff
@@ -0,0 +1,22 @@
+  },
+]
+
+export function statusTag(statusId: LostBed['status'], viewBedspacePage?: boolean) {
```
Maybe change `viewBedspacePage: boolean` to something like `context: 'bookingsAndVoid' | 'voidsOnly'`, so this util doesn't have specific knowledge of a particular page, and give some description as to why there's this variation

bryangaledxw commented on cypress_shared/pages/temporary-accommodation/manage/bedspaceShow.ts:
```diff
@@ -80,4 +101,8 @@ export default class BedspaceShowPage extends Page {
   }
+
+  clickLostBedLink(premisesId: string, roomId: string, lostBedId: string): void {
+    cy.get(`[href="${paths.lostBeds.show({ premisesId, roomId, lostBedId })}"]`).click()
```
Consider changing so it's navigating by the link text? I feel like part of the test is "does the text as the user sees it go to the right place"

bryangaledxw replied:
Ah, there might not be a nice way of doing that, right, since you don't have a CRN like with bookings?

patrickjfl replied:
yeah, exactly

bryangaledxw commented on e2e/tests/stepDefinitions/manage/lostBed.ts:
```diff
@@ -0,0 +1,56 @@
+import lostBedFactory from '../../../../server/testutils/factories/lostBed'
+import LostBedShowPage from '../../../../cypress_shared/pages/temporary-accommodation/manage/lostBedShow'
+
+Given('I mark a bedspace as void', () => {
```
Sorry, realised when I was changing E2E tests myself I was talking nonsense about this! Think you had this right the first time as this doesn't actually mark the bedspace as void, it just goes to the page to do that. There's a rough convention of "I'm X-ing" meaning "I going to the page to do X but not actually doing X yet"

patrickjfl replied:
aha ok! have reverted that one (but not the second, for the reasons you give)

bryangaledxw reviewed (approved):
Looks good, and thanks for the changes! A few more comments, but mostly at the level of take-it-or-leave-it. Your call whether my suggestions would make the code better or not. There's still the stray change to integration tests in the E2E commit, and that commit message that references the type guard util, but I think this is basically ready to go in. Good stuff!

Commit:
Sort imports

Commit:
Extend bookingService

We refactor bookingService to sort bookings and lostBeds into a single table to be rendered to the user on the show bedspace screen.

Commit:
Add E2E testing for lost beds journey
